File: 2024/Python/Challenge#13.py

def is_robot_back(moves: str) -> bool | list[int]:
    i = 0
    double = False
    moves_made = set()
    position = [0, 0]
    moves_list = list(moves)
    
    while i < len(moves_list):
        move = moves_list[i]
        
        normal_moves = {
            'R': [1, 0],
            'L': [-1, 0],
            'U': [0, 1],
            'D': [0, -1]
        }
        
        inverted_moves = {
            'R': 'L',
            'L': 'R',
            'U': 'D',
            'D': 'U'
        }
        
        if move in normal_moves.keys():
            multiplier = 2 if double else 1
            position[0] += normal_moves[move][0] * multiplier
            position[1] += normal_moves[move][1] * multiplier
            double = False
        elif move == '*':
            double = True
        elif move == '!':
            moves_list[i+1] = inverted_moves[moves_list[i+1]]
        elif move == '?':
            if moves_list[i+1] in moves_made:
                i += 1

        moves_made.add(move)
        i += 1
    
    if position[0] == 0 and position[1] == 0:
        return True

    return position


print(is_robot_back('R'))     # [1, 0]
print(is_robot_back('RL'))    # true
print(is_robot_back('RLUD'))  # true
print(is_robot_back('*RU'))   # [2, 1]
print(is_robot_back('R*U'))   # [1, 2]
print(is_robot_back('LLL!R')) # [-4, 0]
print(is_robot_back('R?R'))   # [1, 0]
print(is_robot_back('U?D'))   # true
print(is_robot_back('R!L'))   # [2,0]
print(is_robot_back('U!D'))   # [0,2]
print(is_robot_back('R?L'))   # true
print(is_robot_back('U?U'))   # [0,1]
print(is_robot_back('*U?U'))  # [0,2]
print(is_robot_back('U?D?U')) # true

# Step-by-step examples:
print(is_robot_back('R!U?U')) # [1,0]
# 'R'  -> moves to the right
# '!U' -> inverts and becomes 'D'
# '?U' -> moves upwards, because the 'U' movement hasn't been done yet

print(is_robot_back('UU!U?D')) # [0,1]
# 'U'  -> moves upwards
# 'U'  -> moves upwards
# '!U' -> inverts and becomes 'D'
# '?D' -> does not move, since the 'D' movement has already been done

File: 2024/TypeScript/Challenge#13.ts

function isRobotBack(moves: string): true | [number, number] {
    let i: number = 0;
    let double: boolean = false;
    const movesMade = new Set<string>();
    const position: [number, number] = [0, 0];
    const movesList: string[] = moves.split('');

    const normalMoves: Map<string, [number, number]> = new Map([
        ['R', [1, 0]],
        ['L', [-1, 0]],
        ['U', [0, 1]],
        ['D', [0, -1]]
    ]);

    const invertedMoves: Map<string, string> = new Map([
        ['R', 'L'],
        ['L', 'R'],
        ['U', 'D'],
        ['D', 'U']
    ]);

    while (i < movesList.length) {
        const move = movesList[i];

        if (normalMoves.has(move)) {
            const multiplier = double ? 2 : 1;
            const [x, y] = normalMoves.get(move)!;
            position[0] += x * multiplier;
            position[1] += y * multiplier;
            double = false;
        } else if (move === '*') {
            double = true;
        } else if (move === '!') {
            movesList[i + 1] = invertedMoves.get(movesList[i + 1])!;
        } else if (move === '?') {
            if (movesMade.has(movesList[i + 1])) {
                i += 1;
            }
        }

        movesMade.add(move);
        i += 1;
    }

    return position[0] === 0 && position[1] === 0 ? true : position;
}


console.log(isRobotBack('R'))     // [1, 0]
console.log(isRobotBack('RL'))    // true
console.log(isRobotBack('RLUD'))  // true
console.log(isRobotBack('*RU'))   // [2, 1]
console.log(isRobotBack('R*U'))   // [1, 2]
console.log(isRobotBack('LLL!R')) // [-4, 0]
console.log(isRobotBack('R?R'))   // [1, 0]
console.log(isRobotBack('U?D'))   // true
console.log(isRobotBack('R!L'))   // [2,0]
console.log(isRobotBack('U!D'))   // [0,2]
console.log(isRobotBack('R?L'))   // true
console.log(isRobotBack('U?U'))   // [0,1]
console.log(isRobotBack('*U?U'))  // [0,2]
console.log(isRobotBack('U?D?U')) // true

// Step-by-step examples:
console.log(isRobotBack('R!U?U')) // [1,0]
// 'R'  -> moves to the right 
// '!U' -> inverts and becomes 'D'
// '?U' -> moves upwards, because the 'U' movement hasn't been done yet

console.log(isRobotBack('UU!U?D')) // [0,1]
// 'U'  -> moves upwards
// 'U'  -> moves upwards
// '!U' -> inverts and becomes 'D'
// '?D' -> does not move, since the 'D' movement has already been done
